systemd: allow all systemd services to check selinux status and read efivars

keszybz opened the pull request:
After https://github.com/systemd/systemd/commit/fd5e402fa9 most systemd
services fail to start with:

Oct 27 13:50:38 workstation-uefi systemd[1]: Starting systemd-hostnamed.service...
Oct 27 13:50:38 workstation-uefi systemd-hostnamed[944]: Failed to open SELinux status page: Permission denied
Oct 27 13:50:38 workstation-uefi systemd[1]: systemd-hostnamed.service: Main process exited, code=exited, status=1/FAILURE

After disabling dontaudit:

Oct 27 14:05:08 workstation-uefi audit[1043]: AVC avc:  denied  { read } for  pid=1043 comm="systemd-hostnam" name="status" dev="selinuxfs" ino=19 scontext=system_u:system_r:systemd_hostnamed_t:s0 tcontext=system_u:object_r:security_t:s0 tclass=file permissive=1
Oct 27 14:05:08 workstation-uefi audit[1043]: AVC avc:  denied  { open } for  pid=1043 comm="systemd-hostnam" path="/sys/fs/selinux/status" dev="selinuxfs" ino=19 scontext=system_u:system_r:systemd_hostnamed_t:s0 tcontext=system_u:object_r:security_t:s0 tclass=file permissive=1
Oct 27 14:05:08 workstation-uefi audit[1043]: AVC avc:  denied  { map } for  pid=1043 comm="systemd-hostnam" path="/sys/fs/selinux/status" dev="selinuxfs" ino=19 scontext=system_u:system_r:systemd_hostnamed_t:s0 tcontext=system_u:object_r:security_t:s0 tclass=file permissive=1

This is fully untested, but it seems it might work if
selinux_get_enforce_mode() does what the name suggests.

The check for selinux status is called from mac_selinux_init() which
is called in 16 different places, so I don't think it makes sense to
try to list them all. Similarly, the check for SystemdOptions in
efivarfs is called from shared code, so it doesn't make sense to play
the whack'a'mole game of trying to list them all.

keszybz commented:
Oh wait, https://github.com/SELinuxProject/refpolicy/pull/308/ does this better. But we still don't have this in Fedora :(

zpytela commented:
At a first glance, it seems to me this PR wants to allow all systemd domains read efivarfs, however the first comment and the subsequent link to refpolicy is about accessing security_t. I need to go through all the information first.

keszybz commented:
I guess the #308 PR from refpolicy might be better. It was a while since we merged this PR in systemd and I forgot that a policy PR was also submitted. I'll also file a PR for systemd to fall back to the netlink based API.

keszybz commented:
We also need
```
allow systemd_resolved_t systemd_resolved_var_run_t:sock_file { create unlink };
```

keszybz commented:
```
allow systemd_timedated_t systemd_resolved_t:unix_stream_socket connectto;
```

keszybz committed:
systemd: allow all systemd services to check selinux status

After https://github.com/systemd/systemd/commit/fd5e402fa9 most systemd
services fail to start with:

Oct 27 13:50:38 workstation-uefi systemd[1]: Starting systemd-hostnamed.service...
Oct 27 13:50:38 workstation-uefi systemd-hostnamed[944]: Failed to open SELinux status page: Permission denied
Oct 27 13:50:38 workstation-uefi systemd[1]: systemd-hostnamed.service: Main process exited, code=exited, status=1/FAILURE

After disabling dontaudit:

Oct 27 14:05:08 workstation-uefi audit[1043]: AVC avc:  denied  { read } for  pid=1043 comm="systemd-hostnam" name="status" dev="selinuxfs" ino=19 scontext=system_u:system_r:systemd_hostnamed_t:s0 tcontext=system_u:object_r:security_t:s0 tclass=file permissive=1
Oct 27 14:05:08 workstation-uefi audit[1043]: AVC avc:  denied  { open } for  pid=1043 comm="systemd-hostnam" path="/sys/fs/selinux/status" dev="selinuxfs" ino=19 scontext=system_u:system_r:systemd_hostnamed_t:s0 tcontext=system_u:object_r:security_t:s0 tclass=file permissive=1
Oct 27 14:05:08 workstation-uefi audit[1043]: AVC avc:  denied  { map } for  pid=1043 comm="systemd-hostnam" path="/sys/fs/selinux/status" dev="selinuxfs" ino=19 scontext=system_u:system_r:systemd_hostnamed_t:s0 tcontext=system_u:object_r:security_t:s0 tclass=file permissive=1

As first step, allow all systemd services to check selinux status.
The check for selinux status is called from mac_selinux_init() which
is called in 16 different places, so I don't think it makes sense to
try to list them all. Any code which wants to create a labelled file is
likely to call mac_selinux_init().

keszybz committed:
selinux: tweak selinux_get_enforce_mode() to allow status page to be used

https://github.com/SELinuxProject/refpolicy/pull/308/commits/24827d807379302f5e6d31389911292d304136a1
creates a separate interface for this, but I don't think it makes sense
to complicate things so much. The intent of selinux_get_enforce_mode()
is allow status to be checked. Over time various code that checks it
will transistion to the new more efficient interface. In the meantime, allow
both. At some point in the future, selinux_get_enforce_mode() can be
reduced to
        dev_search_sysfs($1)
	allow $1 security_t:dir list_dir_perms;
	allow $1 security_t:file mmap_read_file_perms;
like selinux_use_status_page() in refpolicy.

keszybz committed:
Allow resolved to created varlink sockets and the domain to talk to it

resolved creates /run/systemd/resolve/io.systemd.Resolve for varlink
communication.

keszybz committed:
Add ability to read init_var_run_t where fs_read_efivarfs_files is allowed

systemd creates a copy of the efivars files in /run to avoid
triggering the misimplemented kernel ratelimit on reads from
/sys/fs/efi. See https://github.com/systemd/systemd/commit/209b2592ed.
This means that any code which used to read the SystemdOptions efivar in the
past will now try to read the file in /run/systemd/efivars/ first, which is
labeled init_var_run_t:

$ ls -lZ /run/systemd/efivars/SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67
-rw-r--r--. 1 root root system_u:object_r:init_var_run_t:s0 67 Oct 28 10:34 /run/systemd/efivars/SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67

Without this patch:
Oct 28 10:34:21 workstation-uefi audit[435]: AVC avc:  denied  { read } for  pid=435 comm="systemd-gpt-aut" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_gpt_generator_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:21 workstation-uefi audit[435]: AVC avc:  denied  { read } for  pid=435 comm="systemd-gpt-aut" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_gpt_generator_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:22 workstation-uefi audit[463]: AVC avc:  denied  { read } for  pid=463 comm="systemd-sysctl" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_sysctl_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:22 workstation-uefi audit[457]: AVC avc:  denied  { read } for  pid=457 comm="systemd-journal" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:syslogd_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:22 workstation-uefi audit[457]: AVC avc:  denied  { read } for  pid=457 comm="systemd-journal" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:syslogd_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:24 workstation-uefi audit[571]: AVC avc:  denied  { read } for  pid=571 comm="systemd-timesyn" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_timedated_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:25 workstation-uefi audit[599]: AVC avc:  denied  { read } for  pid=599 comm="systemd-logind" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_logind_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:26 workstation-uefi audit[601]: AVC avc:  denied  { read } for  pid=601 comm="systemd-network" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_networkd_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:26 workstation-uefi audit[613]: AVC avc:  denied  { read } for  pid=613 comm="systemd-resolve" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_resolved_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:26 workstation-uefi audit[616]: AVC avc:  denied  { read } for  pid=616 comm="systemd-hostnam" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_hostnamed_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:29 workstation-uefi audit[682]: AVC avc:  denied  { read } for  pid=682 comm="systemd-userdbd" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_userdbd_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:29 workstation-uefi audit[684]: AVC avc:  denied  { read } for  pid=684 comm="systemd-userwor" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_userdbd_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:29 workstation-uefi audit[683]: AVC avc:  denied  { read } for  pid=683 comm="systemd-userwor" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_userdbd_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:29 workstation-uefi audit[685]: AVC avc:  denied  { read } for  pid=685 comm="systemd-userwor" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_userdbd_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0
Oct 28 10:34:29 workstation-uefi audit[686]: AVC avc:  denied  { read } for  pid=686 comm="systemd-user-ru" name="SystemdOptions-8cf2644b-4b0b-428f-9387-6d876050dc67" dev="tmpfs" ino=524 scontext=system_u:system_r:systemd_logind_t:s0 tcontext=system_u:object_r:init_var_run_t:s0 tclass=file permissive=0

keszybz committed:
systemd: allow networkd to check namespaces

Oct 28 11:03:49 workstation-uefi audit[597]: AVC avc:  denied  { read open } for  pid=597 comm="systemd-network" path="net:[4026531992]" dev="nsfs" ino=4026531992 scontext=system_u:system_r:systemd_networkd_t:s0 tcontext=system_u:object_r:nsfs_t:s0 tclass=file permissive=0
Oct 28 11:03:49 workstation-uefi systemd-networkd[597]: sd-device-monitor: Unable to get network namespace of udev netlink socket, unable to determine if we are in host netns: Permission denied

keszybz commented:
OK, I think this is ready for merging now. I reworked the commits with much more detailed commit messages and smaller steps. This is blocking systemd-247-rc1 in rawhide, so I'd appreciate a quick review.

With those patches, my rawhide vm boots with only one avc from sssd, which doesn't seem to cause any bigger problems.

keszybz commented:
Ping please.

zpytela commented on policy/modules/kernel/filesystem.if:
```diff
@@ -6700,6 +6700,7 @@ interface(`fs_read_efivarfs_files',`
         ')
 
         read_files_pattern($1, efivarfs_t, efivarfs_t)
+        read_files_pattern($1, init_var_run_t, init_var_run_t)
```
This interface name is fs_read_efivarfs_files(), it should not contain other references.

keszybz replied:
OK, should I add the pattern to read `init_var_run_t` in appropriate places instead?

zpytela replied:
Interfaces are used when a type from another module is needed, supposedly `init_read_pid_files()`. I am also questioning the need as in the current policy version only 2 domains apart from systemd itself have read access to fhe files and a bunch to the directory:
```
# sesearch -A -t init_var_run_t -c file -p read |grep systemd
allow systemd_modules_load_t init_var_run_t:file { getattr ioctl lock open read };
allow systemd_tmpfiles_t non_auth_file_type:file { append create getattr ioctl link lock open read relabelfrom relabelto rename setattr unlink write };

# sesearch -A -t init_var_run_t -c dir |grep systemd
allow systemd_importd_t init_var_run_t:dir { add_name ioctl lock read remove_name write };
allow systemd_logind_t init_var_run_t:dir { add_name ioctl lock read remove_name write };
allow systemd_machined_t init_var_run_t:dir { add_name ioctl lock read remove_name write };
allow systemd_modules_load_t init_var_run_t:dir { ioctl lock read };
allow systemd_notify_t init_var_run_t:dir { add_name ioctl lock read remove_name write };
allow systemd_passwd_agent_t init_var_run_t:dir { add_name create ioctl lock read remove_name write };
allow systemd_resolved_t init_var_run_t:dir { add_name ioctl lock read remove_name write };
allow systemd_timedated_t init_var_run_t:dir { add_name ioctl lock read remove_name write };
allow systemd_tmpfiles_t non_auth_file_type:dir { add_name create getattr ioctl link lock open read relabelfrom relabelto remove_name rename reparent rmdir search setattr unlink write };
allow systemd_userdbd_t init_var_run_t:dir { add_name ioctl lock read remove_name write };
```
Also note some systemd domains use private type for files/dirs in the runtime directory.

zpytela commented on policy/modules/system/systemd.if:
```diff
@@ -22,6 +22,8 @@ template(`systemd_domain_template',`
 
     kernel_read_system_state($1_t)
+    selinux_get_enforce_mode($1_t)
+    systemd_read_efivarfs($1_t)
```
Are both the lines really required for all systemd services/domains to justify having it in the template? I understand it can simplify updates in the future.

keszybz replied:
As I wrote in the commit message, the code path to read the efivar is in low-level code, so anything that looks at kernel options will hit it. In rawhide:
```console
$ journalctl -b-4 --grep avc.*SystemdOpt -o export|grep _COMM|sort -u
_COMM=systemd-gpt-aut
_COMM=systemd-hostnam
_COMM=systemd-journal
_COMM=systemd-logind
_COMM=systemd-network
_COMM=systemd-resolve
_COMM=systemd-sysctl
_COMM=systemd-timesyn
_COMM=systemd-userdbd
_COMM=systemd-user-ru
_COMM=systemd-userwor
```
So yeah, I think it's reasonable to provide blanket approval.

zpytela replied:
It is still less than half of the systemd domains.
Anyway, if it needed to be allowed for every systemd domain, it would be worth considering to remove `read_efivarfs_type` attribute and allow the permissions to systemd_domain instead - but it currently is just a thought.

```
# seinfo -xasystemd_domain

Type Attributes: 1
   attribute systemd_domain;
        systemd_bootchart_t
        systemd_coredump_t
        systemd_gpt_generator_t
        systemd_hostnamed_t
        systemd_hwdb_t
        systemd_importd_t
        systemd_initctl_t
        systemd_localed_t
        systemd_logger_t
        systemd_logind_t
        systemd_machined_t
        systemd_modules_load_t
        systemd_networkd_t
        systemd_notify_t
        systemd_passwd_agent_t
        systemd_resolved_t
        systemd_rfkill_t
        systemd_sysctl_t
        systemd_timedated_t
        systemd_tmpfiles_t
        systemd_userdbd_t

# seinfo -xasystemd_read_efivarfs_type

Type Attributes: 1
   attribute systemd_read_efivarfs_type;
        systemd_gpt_generator_t
        systemd_hostnamed_t
        systemd_hwdb_t
        systemd_localed_t
        systemd_logind_t
        systemd_machined_t
        systemd_modules_load_t
        systemd_networkd_t
        systemd_resolved_t
        systemd_rfkill_t
        systemd_sysctl_t
        systemd_timedated_t
        systemd_userdbd_t
```

zpytela commented on policy/modules/system/systemd.if:
```diff
@@ -2290,16 +2292,16 @@ interface(`systemd_map_resolved_exec_files',`
 ## </param>
 #
-interface(`systemd_dbus_chat_resolved',`
+interface(`systemd_chat_resolved',`
```
All interfaces in the policy currently have "dbus_chat" as a part of the name.

keszybz replied:
Yeah, but now it's either dbus or varlink. How should it be called in that case?

zpytela replied:
I see, so just need to adjust to it.

zpytela commented on policy/modules/system/systemd.te:
```diff
@@ -469,6 +467,7 @@ allow systemd_networkd_t self:packet_socket create_socket_perms;
 allow systemd_networkd_t self:udp_socket create_socket_perms;
 allow systemd_networkd_t self:rawip_socket create_socket_perms;
 allow systemd_networkd_t self:tun_socket { relabelfrom relabelto create_socket_perms };
+allow systemd_networkd_t nsfs_t:file { open read };
```
There is the fs_read_nsfs_files() if to allow the access.

zpytela commented on policy/modules/system/systemd.if:
```diff
@@ -2308,6 +2310,8 @@ interface(`systemd_dbus_chat_resolved',`
 	allow systemd_resolved_t $1:dbus send_msg;
 	ps_process_pattern(systemd_resolved_t, $1)
+
+	allow $1 systemd_resolved_t:unix_stream_socket connectto;
```
This line should rather be placed close to the other allow rules.

zpytela commented on policy/modules/system/systemd.if:
```diff
@@ -2290,16 +2292,16 @@ interface(`systemd_map_resolved_exec_files',`
-##	Send and receive messages from
-##	systemd resolved over dbus.
+##	Exchange messages with
+##	systemd resolved over dbus or varlink.
```
This is just a question: Are current rules sufficient to use varlink?

keszybz replied:
If "current" means "this patch", then yeah, it seems to work OK.

keszybz commented:
Updated based on your comments.

keszybz commented:
OK, I dropped the commit that you didn't like. It's not crucial, because any failure to read efivars is ignored. I hope we can merge this to get systems successfully booting, and then we can work out other details later.

zpytela commented on policy/modules/system/logging.te:
```diff
@@ -607,9 +607,10 @@ files_read_kernel_symbol_table(syslogd_t)
-fs_read_efivarfs_files(syslogd_t)
 fs_search_auto_mountpoints(syslogd_t)
 fs_search_cgroup_dirs(syslogd_t)
+fs_read_efivarfs_files(syslogd_t)
```
Why has this line moved here? It stops following alphabetical order.

keszybz replied:
It's moved together with the other added lines, because make both are required for the same thing. I can move it back if you prefer.

keszybz replied:
Moved  it up now.

zpytela replied:
Thanks. It is not nitpicking, it helps all people going through the rules. It is particularly more important in complex modules.

zpytela commented on policy/modules/system/logging.te:
```diff
@@ -607,9 +607,10 @@ files_read_kernel_symbol_table(syslogd_t)
 fs_search_auto_mountpoints(syslogd_t)
 fs_search_cgroup_dirs(syslogd_t)
+fs_read_efivarfs_files(syslogd_t)
+read_files_pattern(syslogd_t, init_var_run_t, init_var_run_t)
```
There is an interface for accessing types outside the module as mentioned previously.

keszybz replied:
I don't understand what that means. If it should be written in a different way, just say how.

keszybz replied:
`init_read_pid_files(syslogd_t)` — is that it?

zpytela replied:
Yes, I elaborated on that in the review a week ago.

zpytela commented on policy/modules/system/systemd.if:
```diff
@@ -22,6 +22,7 @@ template(`systemd_domain_template',`
     init_nnp_daemon_domain($1_t)
 
     kernel_read_system_state($1_t)
+    selinux_get_enforce_mode($1_t)
```
Move the line below, kernel is the only exception to alphabetical order when using external interfaces.

zpytela commented on policy/modules/system/systemd.te:
```diff
@@ -13,6 +13,7 @@ attribute systemd_private_tmp_type;
 
 attribute systemd_read_efivarfs_type;
 fs_read_efivarfs_files(systemd_read_efivarfs_type)
+read_files_pattern(systemd_read_efivarfs_type, init_var_run_t, init_var_run_t)
```
I see this questionable, too. Is the only reason the `/run/systemd/efivars/SystemdOptions-.*` file? Should we rather label the directory with a private type?

The `init_var_run_t` is assigned to many objects in /run/systemd which do not have their own type. With this rule, the access to generic systemd data is allowed to 14 systemd domains. Is this desirable?

(Another problem is overloading the "efivarfs" attribute with another meaning, but this is not easy to solve.)

Anyway, if there is the need, use an interface and move it below to local policy rules section.

keszybz replied:
I consider this a temporary workaround. The cache for efivars in /run will hopefully go away as soon as the kernel ratelimit is fixed to take multiple readers into account. `init_var_run_t` is publicly readable data, so I don't think it makes sense to add a new type and make things more complicated for something that might go away next month.

zpytela replied:
I understand. However, with this broad access some existing bugzillas can be made resolved, so we are about to introduce a planned regression.

keszybz replied:
Yeah, but it's all public data. Both the stuff that was already under this label, and the "new" efivars stuff. It's publicly readable in the fs under /run, but a lot of it is also available over dbus to all unprivileged users. I'm not sure why that would be a "planned regression" — at the point where systemd stops providing the efivars cache, this rule can be dropped again and we return to status quo ante.

zpytela replied:
So is this a temporary solution or a legit request? Note there are also objects with a different type. I was talking about regression as it looked you wanted to revert it back later.
```
# ls -Zago /run/systemd/
total 0
drwxr-xr-x. 21 system_u:object_r:init_var_run_t:s0                    480 Nov 12 11:56 .
drwxr-xr-x. 27 system_u:object_r:var_run_t:s0                         780 Nov 12 11:56 ..
drwxr-xr-x.  2 system_u:object_r:systemd_passwd_var_run_t:s0           40 Nov 12 11:56 ask-password
srw-------.  1 system_u:object_r:init_var_run_t:s0                      0 Nov 12 11:56 coredump
drwxr-xr-x.  7 system_u:object_r:systemd_unit_file_t:s0               300 Nov 12 11:56 generator
drwxr-xr-x.  3 system_u:object_r:init_var_run_t:s0                     60 Nov 12 11:56 generator.early
drwxr-xr-x.  2 system_u:object_r:init_var_run_t:s0                     60 Nov 12 11:56 home
drwxr-xr-x.  3 system_u:object_r:init_var_run_t:s0                    160 Nov 12 11:56 inaccessible
drwxr-xr-x.  2 system_u:object_r:systemd_logind_inhibit_var_run_t:s0   80 Nov 12 11:56 inhibit
drwxr-xr-x.  3 system_u:object_r:syslogd_var_run_t:s0                 180 Nov 12 11:56 journal
drwxr-xr-x.  2 system_u:object_r:systemd_machined_var_run_t:s0         40 Nov 12 11:56 machines
drwxr-xr-x.  5 system_u:object_r:systemd_networkd_var_run_t:s0        100 Nov 12 11:56 netif
srwxrwxrwx.  1 system_u:object_r:init_var_run_t:s0                      0 Nov 12 11:56 notify
srwxrwxrwx.  1 system_u:object_r:init_var_run_t:s0                      0 Nov 12 11:56 private
drwxr-xr-x.  3 system_u:object_r:systemd_resolved_var_run_t:s0        100 Nov 12 11:58 resolve
drwxr-xr-x.  2 system_u:object_r:systemd_logind_var_run_t:s0           60 Nov 12 11:56 seats
drwxr-xr-x.  2 system_u:object_r:systemd_logind_sessions_t:s0          80 Nov 12 11:57 sessions
drwxr-xr-x.  2 system_u:object_r:systemd_logind_var_run_t:s0           40 Nov 12 11:56 shutdown
drwxr-xr-x.  2 system_u:object_r:systemd_unit_file_t:s0                40 Nov 12 11:56 system
drwxr-xr-x.  2 system_u:object_r:init_var_run_t:s0                     60 Nov 12 11:57 transient
drwx------.  2 system_u:object_r:init_var_run_t:s0                     40 Nov 12 11:56 unit-root
drwxr-xr-x.  2 system_u:object_r:systemd_unit_file_t:s0              1080 Nov 12 11:58 units
drwxr-xr-x.  2 system_u:object_r:systemd_userdbd_runtime_t:s0         120 Nov 12 11:56 userdb
drwxr-xr-x.  2 system_u:object_r:systemd_logind_var_run_t:s0           60 Nov 12 11:57 users
```

keszybz replied:
> So is this a temporary solution or a legit request?

Both? It's temporary in the sense that it hopefully will not be necessary forever. But there are no plans that I know of to change this in the kernel, so it'll be necessary at least until current and upcoming kernels go out of scope, i.e. a long while.

> Note there are also objects with a different type.

I don't understand the significance of objects with a different type, since this rule doesn't matter for them at all. 

So... I think the current rule is OK because it a) solves the issue at hand, b) it does allow access to more stuff than strictly necessary, but that stuff is public anyway so it's OK to allow access to it.

zpytela replied:
@keszybz thank you for your effort to prepare this complex set of patches. I will merge this PR now.

Note removing rules from policy is troublesome as it can have unpredictable effect, depending on changes in other components requirement in the near future.

zpytela commented:
@keszybz requesting now minor changes only, plus the typo "netword" in 5th commit (I dislike typos only in a few cases).

With the access to `init_var_run_t`, I'd need deeper systemd understanding, so rather questioning.

keszybz commented:
Another version. It still allows access to `init_var_run_t` fairly widely, which I think is reasonable. I think I addressed other comments.

With this I get some avcs:
```
Nov 09 19:05:55 fedora audit[596]: AVC avc:  denied  { write } for  pid=596 comm="chronyd" name="io.systemd.Resolve" dev="tmpfs" ino=956 scontext=system_u:system_r:chronyd_t:s0 tcontext=system_u:object_r:systemd_resolved_var_run_t:s0 tclass=sock_file permissive=0
Nov 09 19:05:55 fedora audit[597]: AVC avc:  denied  { write } for  pid=597 comm="rsyslogd" name="io.systemd.Resolve" dev="tmpfs" ino=956 scontext=system_u:system_r:syslogd_t:s0 tcontext=system_u:object_r:systemd_resolved_var_run_t:s0 tclass=sock_file permissive=0
Nov 09 19:05:59 fedora audit[740]: AVC avc:  denied  { write } for  pid=740 comm="hostname" name="io.systemd.Resolve" dev="tmpfs" ino=956 scontext=system_u:system_r:cockpit_ws_t:s0 tcontext=system_u:object_r:systemd_resolved_var_run_t:s0 tclass=sock_file permissive=0
Nov 09 19:05:59 fedora audit[596]: AVC avc:  denied  { write } for  pid=596 comm="chronyd" name="io.systemd.Resolve" dev="tmpfs" ino=956 scontext=system_u:system_r:chronyd_t:s0 tcontext=system_u:object_r:systemd_resolved_var_run_t:s0 tclass=sock_file permissive=0
```
but I don't think those would have occurred with the previous versions of the PR too, but I didn't see them for some reason (maybe they were blocked by one of the earlier denials?). I'm not sure how to handle those. Essentially, we want everyone to be able to access `io.systemd.Resolve`.

keszybz commented:
We tagged v247-rc2 today, and the rawhide build is still blocked on this.

zpytela commented:
There seems to be last ambiguity regarding
```
read_files_pattern(systemd_read_efivarfs_type, init_var_run_t, init_var_run_t)
```
Once clarified, I'll create a rawhide build.
